apps/emqx_bridge_greptimedb/src/emqx_bridge_greptimedb_connector.erl

```diff
@@ -0,0 +1,627 @@
+    Config = #{
+        server := Server
+    }
+) ->
+    #{hostname := Host, port := Port} = emqx_schema:parse_server(Server, ?GREPTIMEDB_HOST_OPTIONS),
+    [
+        {endpoints, [{http, str(Host), Port}]},
+        {pool_size, erlang:system_info(schedulers)},
+        {pool, InstId},
+        {pool_type, random},
+        {timeunit, maps:get(precision, Config, ms)}
+    ] ++ protocol_config(Config).
+
+protocol_config(
+    #{
+        dbname := DbName,
+        ssl := SSL
+    } = Config
+) ->
+    [
+        {dbname, DbName}
+    ] ++ auth(Config) ++
+        ssl_config(SSL).
+
+ssl_config(#{enable := false}) ->
+    [
+        {https_enabled, false}
+    ];
+ssl_config(SSL = #{enable := true}) ->
+    [
+        {https_enabled, true},
+        {transport, ssl},
+        {transport_opts, emqx_tls_lib:to_client_opts(SSL)}
+    ].
+
+auth(#{username := Username, password := Password}) ->
+    [
+        {auth, {basic, #{username => Username, password => Password}}}
+    ];
+auth(_) ->
+    [].
+
+redact_auth(Term) ->
+    emqx_utils:redact(Term, fun is_auth_key/1).
+
+is_auth_key(Key) when is_binary(Key) ->
+    string:equal("authorization", Key, true);
+is_auth_key(_) ->
+    false.
+
+%% -------------------------------------------------------------------------------------------------
+%% Query
+do_query(InstId, Client, Points) ->
+    case greptimedb:write_batch(Client, Points) of
+        {ok, #{response := {affected_rows, #{value := Rows}}}} ->
+            ?SLOG(debug, #{
+                msg => "greptimedb write point success",
+                connector => InstId,
+                points => Points
+            }),
+            {ok, {affected_rows, Rows}};
+        {error, {unauth, _, _}} ->
+            ?tp(greptimedb_connector_do_query_failure, #{error => <<"authorization failure">>}),
+            ?SLOG(error, #{
+                msg => "greptimedb_authorization_failed",
+                client => redact_auth(Client),
+                connector => InstId
+            }),
+            {error, {unrecoverable_error, <<"authorization failure">>}};
+        {error, Reason} = Err ->
+            ?tp(greptimedb_connector_do_query_failure, #{error => Reason}),
+            ?SLOG(error, #{
+                msg => "greptimedb write point failed",
+                connector => InstId,
+                reason => Reason
+            }),
+            case is_unrecoverable_error(Err) of
+                true ->
+                    {error, {unrecoverable_error, Reason}};
+                false ->
+                    {error, {recoverable_error, Reason}}
+            end
+    end.
+
+%% -------------------------------------------------------------------------------------------------
+%% Tags & Fields Config Trans
+
+to_config(Lines, Precision) ->
+    to_config(Lines, [], Precision).
+
+to_config([], Acc, _Precision) ->
+    lists:reverse(Acc);
+to_config([Item0 | Rest], Acc, Precision) ->
+    Ts0 = maps:get(timestamp, Item0, undefined),
```

Comment:
nit: the `undefined` could be replaced with the `?DEFAULT_TIMESTAMP_TMPL`, so the first clause of `preproc_tmpl_timestamp` can be deleted

Reply:
Some code is from the influxdb data bridge, do I need to modify it too?

Reply:
This is just small ideas, feel free to decide, but I think we don't need to consider the influxdb, it outside this PR

Reply:
I changed the `undefined` to `?DEFAULT_TIMESTAMP_TMPL` in this line. But the first clause of `preproc_tmpl_timestamp` can't be deleted, because the value of `timestamp` in `Item0` may be `undefined`, we have to process it.